presto-orc/src/main/java/com/facebook/presto/orc/reader/SliceDirectSelectiveStreamReader.java

```diff
@@ -0,0 +1,537 @@
+    protected int outputPositionCount;
+
+    protected final boolean outputRequired;
+    protected boolean allNulls = true;
+    private boolean[] isNullVector;
+    private int[] lengthVector;
+    private int[] offsetVector;
+    private int totalLength;
+    private boolean valuesInUse;
+    private byte[] data;
+    private Slice slice;
+    private int lengthIndex;
+    private int maxLength;
+    private boolean outputPositionsReadOnly;
+
+    public SliceDirectSelectiveStreamReader(StreamDescriptor streamDescriptor, Optional<TupleDomainFilter> filter, Optional<Type> outputType, LocalMemoryContext newLocalMemoryContext)
+    {
+        this.streamDescriptor = requireNonNull(streamDescriptor, "streamDescriptor is null");
+        this.filter = requireNonNull(filter, "filter is null").orElse(null);
+        this.systemMemoryContext = newLocalMemoryContext;
+        this.nonDeterministicFilter = this.filter != null && !this.filter.isDeterministic();
+        this.nullsAllowed = this.filter == null || nonDeterministicFilter || this.filter.testNull();
+        this.outputType = requireNonNull(outputType, "outputType is null").orElse(null);
+        this.outputRequired = outputType.isPresent();
+        if (!outputRequired) {
+            checkState(filter != null, "filter should be present if output is not required");
+        }
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (filter != null) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
+        }
+        else {
+            outputPositions = positions;
+            outputPositionsReadOnly = true;
+        }
+
+        // account memory used by values, nulls, outputPositions, offsetVector, lengthVector, isNullVector
+        systemMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        if (readOffset < offset) {
+            skip(offset - readOffset);
+        }
+
+        prepareForNextRead(positionCount, positions);
+
+        ensureValuesCapacity(positionCount, presentStream != null && nullsAllowed);
+
+        int streamPosition;
+
+        if (dataStream == null && lengthVector == null && isNullVector != null) {
+            streamPosition = readAllNulls(positions, positionCount);
+        }
+        else if (filter == null) {
+            streamPosition = readNoFilter(positions, positionCount);
+        }
+        else {
+            streamPosition = readWithFilter(positions, positionCount);
+        }
+
+        readOffset = offset + streamPosition;
+        return outputPositionCount;
+    }
+
+    private int readNoFilter(int[] positions, int positionCount)
+            throws IOException
+    {
+        int streamPosition = 0;
+        allNulls = false;
+
+        int offsetVectorIndex = 0;
```

Comment:
- is there any particular reason to include `vector` in the names? I think offsets and offsetIndex are easier to read.

Reply:
no reused form batch reader